### ISSUE TITLE
o.lift(CustomOperator()).toBlocking().single()  is deadlock  after 1.0.9

### ISSUE THREAD

**Issue by neocoin** [ReactiveX/RxJava#3071]
This issue relate https://github.com/ReactiveX/RxJavaString/issues/23  in rxjava-string

This code is working on 1.0.8  but not on 1.0.9 

```java
Observable<String> columns = Observable.from("a,b,c,d".split(","))
        .map(i -> String.format("%s.%s", "prefix", i));
Observable<String> query = StringObservable.join(columns, ",");
System.out.println(query.toBlocking().single());  // deadlock 
```

output before 1.0.9 

```
prefix.a,prefix.b,prefix.c,prefix.d
```

But after some tests, I'm sure this is not relate rxjava-string.

For reproduction, I remove rxjava-string code dependency, add simple custom operation. 
(ref  https://github.com/ReactiveX/RxJava/wiki/Implementing-Your-Own-Operators#sequence-operators )

This custom operator pass only first element. 

```java
Observable<String> one = Observable.from("a,b,c,d".split(",")).lift(new Observable.Operator<String, String>() {
    @Override
    public Subscriber<? super String> call(Subscriber<? super String> subscriber) {
        return new Subscriber(subscriber) {
            boolean runFirst = false;

            @Override
            public void onCompleted() {
                System.out.println("call in onCompleted");
                if(!subscriber.isUnsubscribed()) {
                    subscriber.onCompleted();
                }
            }

            @Override
            public void onError(Throwable throwable) {
                System.out.println(throwable);
            }

            @Override
            public void onNext(Object o) {
                System.out.println(o);
                if(!runFirst) {
                    subscriber.onNext(o.toString());
                    runFirst = true;
                }
            }
        };
    }
});
System.out.println(one.toBlocking().single());
```

working on 1.0.8 but  not on 1.0.9

```
value => a in onNext
value => b in onNext
call in onCompleted
a
```

Deadlock issue on https://github.com/ReactiveX/RxJava/blob/1.x/src/main/java/rx/observables/BlockingObservable.java#L461  line

```java
            latch.await();  // wait for endless
```

Because runtime doesn't reach 'onComplete' in BlockingObservable#blockForSingle. ( https://github.com/ReactiveX/RxJava/blob/1.x/src/main/java/rx/observables/BlockingObservable.java#L445 )

So I test with BlockingObservable.java code from 1.0.12 to 1.0.2 but it's work fine. 

Now, I just guess 'OperatorSingle should request exactly what it needs' (  https://github.com/ReactiveX/RxJava/pull/2860 ) commit's side-effect, but I'm not certain exactly problem point or it is possible from my wrong usage because I don't know OperatorSingle changes history. 

@davidmoten , Could you check this issue?

**Comment by akarnokd** [comment]
Hello and sorry for the late answer. 

The bug is in the request management of `StringObservable.join()` and in your custom operator as well. The problem is that by using the Subscriber constructor, you made a direct connection between the downstream requesting and the emission of the `from` operator which instead of receiving a Long.MAX_VALUE request from join(), it gets a request of 2.

If you change your custom operator to this inside the lifted function, you should be able to see the expected result:

```java
Subscriber<String> s = new Subscriber<String>() {
    boolean runFirst = false;

    @Override
    public void onCompleted() {
        System.out.println("call in onCompleted");
        if(!subscriber.isUnsubscribed()) {
            subscriber.onCompleted();
        }
    }

    @Override
    public void onError(Throwable throwable) {
        System.out.println(throwable);
    }

    @Override
    public void onNext(String o) {
        System.out.println(o);
        if(!runFirst) {
            subscriber.onNext(o.toString());
            runFirst = true;
        }
    }
};
subscriber.add(s);
return s;
```

**Comment by akarnokd** [comment]
I forgot the RxJavaString has been released with version 1.0.1 where the issue has been fixed and your original code works for me. Thanks for the report!
